### PR TITLE
[JENKINS-49675] Mark the cookie as secure when serving via HTTPS

### PR DESCRIPTION
**Pull request by jtnord**
See [JENKINS-49675](https://issues.jenkins-ci.org/browse/JENKINS-49675). (ish)

the screeenresolution cookie was not sent as "secure" when the request was via a HTTPs server.

There is nothing sensitive here - but some scanners flag this as bad assuming all cookies are equal.

If the request is secure (it is via HTTPs or has been reverse proxied with https) then set the secure flag for the cookie.

Tested manually rather than with tests as I did not know a way to use JenkinsRule or the like with https

So used `java -jar war\target\jenkins.war --httpListenAddress=127.0.0.1 --httpsListenAddress=127.0.0.1 --httpPort=8080 --httpsPort=8443` and loggin in and using chrome dev tools to look at the cookies.

Also performed a test with an apache reverse proxy terminating the TLS session (and Jenkins using vanilla http) with https://github.com/jglick/jenkins-demo-reverse-proxy

### Proposed changelog entries

* Screen Resolution cookie now has the `secure` flag set when Jenkins is running on HTTPS.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [x] There are at least 2 approvals for the pull request and no outstanding requests for change
- [x] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [x] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [x] Proper changelog labels are set so that the changelog can be generated automatically
- [x] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/resources/hudson/model/View/index.jelly
Comment (jglick) [inline core/src/main/resources/hudson/model/View/index.jelly:49]: idiomatically
```suggestion
              <j:when test="${req.secure}">
```

File: core/src/main/resources/hudson/model/View/index.jelly
Comment (jglick) [inline core/src/main/resources/hudson/model/View/index.jelly:47]: As a rule, if you find the need to use `j:invokeStatic`, it is a good sign you should be pushing the computation into the controller (`View.java`).

In this case, IIRC `request` or something like it is already bound in page variables.

File: core/src/main/resources/hudson/model/View/index.jelly
Comment (jglick) [inline core/src/main/resources/hudson/model/View/index.jelly:53]: Could we not just have one branch?
```suggestion
                    secure: ${req.secure}
```
or similar?

File: core/src/main/resources/hudson/model/View/index.jelly
Comment (Wadeck) [inline core/src/main/resources/hudson/model/View/index.jelly:47]: I think you could use `request` directly. 

Source: https://github.com/stapler/stapler/blob/58a0fd5d7e336272a3c90140be632248b8f9a45c/jelly/src/main/java/org/kohsuke/stapler/jelly/DefaultScriptInvoker.java#L159

**Review by Wadeck — APPROVED** [review]
Seems fine, not tested manually but the approach is what I would have done 👍

File: core/src/main/resources/hudson/model/View/index.jelly
Comment (jtnord) [reply core/src/main/resources/hudson/model/View/index.jelly:47]: I could.  I went looking for the documentation and did not see it exposed, obviously not looking hard enough :)

File: core/src/main/resources/hudson/model/View/index.jelly
Comment (jtnord) [reply core/src/main/resources/hudson/model/View/index.jelly:53]: we could and I am happy to do that.  I got the impression @Wadeck was not in favour - but as this is just a boolean it should be completely safe

Comment (Wadeck) [reply core/src/main/resources/hudson/model/View/index.jelly:53]: both works, for CSP we will eventually move it to its own file ;)

Comment (jtnord) [reply core/src/main/resources/hudson/model/View/index.jelly:53]: > for CSP we will eventually move it to its own file

in which case we would need 2 files IIUC, or the javascript would need to work out if the endpoint would be https and then do the right thing?

would it be better to keep if choose/when/then  because of that?

happy to change to whatever the consensus is with @jglick

Comment (Wadeck) [reply core/src/main/resources/hudson/model/View/index.jelly:53]: 💡 By using the technique from https://www.jenkins.io/doc/developer/security/xss-prevention/#passing-values-to-javascript, you can pass the variable as an attribute, that will be read by the javascript. With this approach, you can have a single JavaScript file that will behave differently according to the `secure`.
(this is out of the scope for this PR, I don't want to force you doing that just for this secure cookie :D)

Comment (jglick) [reply core/src/main/resources/hudson/model/View/index.jelly:53]: > whatever the consensus is with @jglick

No strong opinion.

**Review by Wadeck — APPROVED** [review]
Thanks James for the YUI investigation, you're the new expert in that library :trollface:

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
